### ISSUE TITLE
Lint for unused empty function value

### ISSUE THREAD

**Issue by Oakchris1955** [rust-lang/rust-clippy#15957]
### What it does

Suggests turning `let _unused = returns_nothing();` to `returns_nothing();`

### Advantage

Increases code hygiene

### Drawbacks

None

### Example

```rust
let _unused: () = returns_nothing();
```

Could be written as:

```rust
returns_nothing();
```

### Comparison with existing lints

No similar lints were found

### Additional Context

A quick [Github search](https://github.com/search?q=%22%3A+%28%29+%3D+%22+language%3ARust&type=code&l=Rust) shows around 116 thousand instances of the aforementioned example

**Comment by samueltardieu** [comment]
This lint exists and is called `let_unit_value`, however it doesn't trigger when the type is explicitly present. It should be fixed.

@rustbot label -A-lint C-bug I-false-negative "good first issue"

**Comment by samueltardieu** [comment]
Note that there is a small subtlety here: the lint should not trigger if the explicit type is needed to force the type of the initialization expression, for example when generics are involved.

**Comment by matanshavit** [comment]
Is this what was intended? [#15975](https://github.com/rust-lang/rust-clippy/pull/15975)

**Comment by Oakchris1955** [comment]
From what I see yes.

**Comment by Jarcho** [comment]
Not linting here is intentional. People use `let _: () = ...` as an assertion that the expression actually evaluates to a unit type.

**Comment by samueltardieu** [comment]
In this case they can use `let () = …` which is more compact while being as explicit. Should we suggest that in a note?

**Comment by Jarcho** [comment]
It should just be the suggestion if we're going to touch this at all. `let () = ...` and `let _: () = ...` are equally clear in intent so we should either lint against both (we've already received pushback on that), or suggest writing one over the other for consistency.

**Comment by Oakchris1955** [comment]
I see no difference between the two. Imo, both should be linted

**Comment by Jarcho** [comment]
`let _some_name: () = ...` should still be linted against, but it should be transformed to `let () = ...`. The discussion on this has already happened (multiple times IIRC) and there's too many people against it to do that by default.
